// parser/index.js
/* parser generated by jison 0.4.18 */
/*
  Returns a Parser object of the following structure:

  Parser: {
    yy: {}
  }

  Parser.prototype: {
    yy: {},
    trace: function(),
    symbols_: {associative list: name ==> number},
    terminals_: {associative list: number ==> name},
    productions_: [...],
    performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate, $$, _$),
    table: [...],
    defaultActions: {...},
    parseError: function(str, hash),
    parse: function(input),

    lexer: {
        EOF: 1,
        parseError: function(str, hash),
        setInput: function(input),
        input: function(),
        unput: function(str),
        more: function(),
        less: function(n),
        pastInput: function(),
        upcomingInput: function(),
        showPosition: function(),
        test_match: function(regex_match_array, rule_index),
        next: function(),
        lex: function(),
        begin: function(condition),
        popState: function(),
        _currentRules: function(),
        topState: function(),
        pushState: function(condition),

        options: {
            ranges: boolean           (optional: true ==> token location info will include a .range[] member)
            flex: boolean             (optional: true ==> flex-like lexing behaviour where the rules are tested exhaustively to find the longest match)
            backtrack_lexer: boolean  (optional: true ==> lexer regexes are tested in order and for each matching regex the action code is invoked; the lexer terminates the scan when a token is returned by the action code)
        },

        performAction: function(yy, yy_, $avoiding_name_collisions, YY_START),
        rules: [...],
        conditions: {associative list: name ==> set},
    }
  }


  token location info (@$, _$, etc.): {
    first_line: n,
    last_line: n,
    first_column: n,
    last_column: n,
    range: [start_number, end_number]       (where the numbers are indexes into the input string, regular zero-based)
  }


  the parseError function receives a 'hash' object with these members for lexer and parser errors: {
    text:        (matched text)
    token:       (the produced terminal token, if any)
    line:        (yylineno)
  }
  while parser (grammar) errors will also provide these members, i.e. parser errors deliver a superset of attributes: {
    loc:         (yylloc)
    expected:    (string describing the set of expected tokens)
    recoverable: (boolean: TRUE when the parser has a error recovery rule available for this particular error)
  }
*/
var index = (function(){
var o=function(k,v,o,l){for(o=o||{},l=k.length;l--;o[k[l]]=v);return o},$V0=[1,5],$V1=[5,7];
var parser = {trace: function trace() { },
yy: {},
symbols_: {"error":2,"program":3,"import_decrs":4,"EOF":5,"import_decr":6,"import":7,"identifier":8,"terminator":9,"type_decrs":10,"type_decr":11,"class_decr":12,"public":13,"class":14,"extend_decr":15,"class_body":16,"extends":17,"set_start":18,"class_body_decrs":19,"set_end":20,"class_body_decr":21,"class_member_decr":22,"consr_declarator":23,"consr_body":24,"field_decr":25,"method_decr":26,"paranthesis_start":27,"formal_parameter_list":28,"paranthesis_end":29,"separator":30,"formal_parameter":31,"type":32,"var_declarator_id":33,"explicit_consr_invocation":34,"block_stmts":35,"this":36,"argument_list":37,"super":38,"var_declarators":39,"var_declarator":40,"op_assign":41,"var_init":42,"brackets_start":43,"brackets_end":44,"expr":45,"array_init":46,"void":47,"method_declarator":48,"method_body":49,"block":50,"var_inits":51,"primitive_type":52,"reference_type":53,"integral_type":54,"floating_type":55,"boolean":56,"byte":57,"short":58,"int":59,"long":60,"char":61,"float":62,"double":63,"block_stmt":64,"stmt":65,"stmt_wots":66,"if_then_stmt":67,"if_then_else_stmt":68,"while_stmt":69,"for_stmt":70,"stmt_nsi":71,"if_then_else_stmt_nsi":72,"while_stmt_nsi":73,"for_stmt_nsi":74,"switch_stmt":75,"do_stmt":76,"break_stmt":77,"continue_stmt":78,"return_stmt":79,"stmt_expr":80,"if":81,"else":82,"switch":83,"switch_block":84,"switch_block_stmt_groups":85,"switch_labels":86,"switch_block_stmt_group":87,"switch_label":88,"case":89,"literal":90,"colon":91,"default":92,"while":93,"do":94,"for":95,"for_init":96,"stmt_expr_list":97,"break":98,"continue":99,"return":100,"cond_expr":101,"assignment":102,"preinc_expr":103,"predec_expr":104,"post_expr":105,"method_invocation":106,"class_instance_creation_expr":107,"left_hand_side":108,"assignment_operator":109,"expr_name":110,"field_access":111,"array_access":112,"op_mulAssign":113,"op_divAssign":114,"op_modAssign":115,"op_addAssign":116,"op_subAssign":117,"op_LshiftEqual":118,"op_RshiftEqual":119,"op_andAssign":120,"op_orAssign":121,"op_xorAssign":122,"cond_or_expr":123,"cond_and_expr":124,"op_oror":125,"incl_or_expr":126,"op_andand":127,"excl_or_expr":128,"op_or":129,"and_expr":130,"op_xor":131,"equality_expr":132,"op_and":133,"relational_expr":134,"op_equalCompare":135,"op_notequalCompare":136,"shift_expr":137,"op_greater":138,"op_greaterEqual":139,"op_less":140,"op_lessEqual":141,"instanceof":142,"additive_expr":143,"op_Lshift":144,"op_Rshift":145,"multiplicative_expr":146,"op_add":147,"op_sub":148,"unary_expr":149,"op_mul":150,"op_div":151,"op_mod":152,"op_decrement":153,"op_increment":154,"sign":155,"unary_expr_npm":156,"postfix_expr":157,"op_not":158,"cast_expr":159,"postdec_expr":160,"postinc_expr":161,"primary":162,"field_invoker":163,"dim_exprs":164,"primary_no_new_array":165,"array_creation_expr":166,"new":167,"dims":168,"dim_expr":169,"integer_literal":170,"float_literal":171,"boolean_literal":172,"character_literal":173,"string_literal":174,"null_literal":175,"$accept":0,"$end":1},
terminals_: {2:"error",5:"EOF",7:"import",8:"identifier",9:"terminator",13:"public",14:"class",17:"extends",18:"set_start",20:"set_end",27:"paranthesis_start",29:"paranthesis_end",30:"separator",36:"this",38:"super",41:"op_assign",43:"brackets_start",44:"brackets_end",47:"void",56:"boolean",57:"byte",58:"short",59:"int",60:"long",61:"char",62:"float",63:"double",81:"if",82:"else",83:"switch",89:"case",91:"colon",92:"default",93:"while",94:"do",95:"for",98:"break",99:"continue",100:"return",113:"op_mulAssign",114:"op_divAssign",115:"op_modAssign",116:"op_addAssign",117:"op_subAssign",118:"op_LshiftEqual",119:"op_RshiftEqual",120:"op_andAssign",121:"op_orAssign",122:"op_xorAssign",125:"op_oror",127:"op_andand",129:"op_or",131:"op_xor",133:"op_and",135:"op_equalCompare",136:"op_notequalCompare",138:"op_greater",139:"op_greaterEqual",140:"op_less",141:"op_lessEqual",142:"instanceof",144:"op_Lshift",145:"op_Rshift",147:"op_add",148:"op_sub",150:"op_mul",151:"op_div",152:"op_mod",153:"op_decrement",154:"op_increment",158:"op_not",163:"field_invoker",167:"new",170:"integer_literal",171:"float_literal",172:"boolean_literal",173:"character_literal",174:"string_literal",175:"null_literal"},
productions_: [0,[3,2],[3,1],[4,1],[4,2],[6,3],[10,2],[10,1],[11,1],[11,1],[12,5],[12,4],[12,4],[12,3],[15,2],[16,3],[19,2],[19,1],[21,1],[21,3],[21,2],[22,1],[22,1],[23,4],[28,3],[28,1],[28,0],[31,2],[24,4],[24,3],[24,3],[24,2],[34,4],[34,4],[34,3],[34,3],[25,4],[25,3],[39,3],[39,1],[40,1],[40,3],[33,3],[33,1],[42,1],[42,1],[26,4],[26,4],[26,3],[26,3],[48,4],[49,1],[46,4],[46,3],[46,3],[46,2],[51,3],[51,1],[32,1],[32,1],[52,1],[52,1],[52,1],[54,1],[54,1],[54,1],[54,1],[54,1],[55,1],[55,1],[53,1],[53,3],[50,3],[50,2],[35,2],[35,1],[64,3],[64,1],[65,1],[65,1],[65,1],[65,1],[65,1],[71,1],[71,1],[71,1],[71,1],[66,1],[66,1],[66,1],[66,1],[66,1],[66,1],[66,2],[66,1],[67,5],[68,7],[72,7],[75,5],[84,4],[84,3],[84,3],[84,2],[85,2],[85,1],[87,2],[86,2],[86,1],[88,3],[88,5],[88,2],[69,5],[73,5],[76,7],[70,9],[70,8],[70,8],[70,8],[70,7],[70,7],[70,7],[70,6],[74,9],[74,8],[74,8],[74,8],[74,7],[74,7],[74,7],[74,6],[96,1],[96,2],[97,3],[97,1],[77,2],[78,2],[79,3],[79,2],[45,1],[45,1],[80,1],[80,1],[80,1],[80,1],[80,1],[80,1],[102,3],[108,1],[108,1],[108,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[109,1],[101,1],[123,1],[123,3],[124,1],[124,3],[126,1],[126,3],[128,1],[128,3],[130,1],[130,3],[132,1],[132,3],[132,3],[134,1],[134,3],[134,3],[134,3],[134,3],[134,3],[137,1],[137,3],[137,3],[143,1],[143,3],[143,3],[146,1],[146,3],[146,3],[146,3],[104,2],[103,2],[149,1],[149,1],[149,2],[149,1],[156,1],[156,1],[156,2],[156,1],[159,4],[160,2],[160,2],[161,2],[161,2],[105,1],[105,1],[157,1],[157,1],[106,4],[106,6],[106,6],[106,3],[106,5],[106,5],[111,3],[111,3],[112,3],[112,3],[162,1],[162,1],[165,1],[165,1],[165,3],[165,1],[165,1],[165,1],[165,1],[107,5],[107,4],[37,1],[37,3],[166,4],[166,4],[166,3],[166,3],[164,2],[164,1],[169,3],[168,2],[168,3],[110,1],[110,3],[90,1],[90,1],[90,1],[90,1],[90,1],[90,1],[155,1],[155,1]],
performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */) {
/* this == yyval */

var $0 = $$.length - 1;
switch (yystate) {
case 1:
 return { nt: 'program', children: [$$[$0-1],{ t: 'EOF', l: $$[$0] }] } 
break;
case 2:
 return { nt: 'program', children: [{ t: 'EOF', l: $$[$0] }] } 
break;
case 3:
 this.$ = { nt: 'import_decrs', children: [$$[$0]] } 
break;
case 4:
 this.$ = { nt: 'import_decrs', children: [$$[$0-1],$$[$0]] } 
break;
case 5:
 this.$ = { nt: 'import_decr', children: [{ t: 'import', l: $$[$0-2] },{ t: 'identifier', l: $$[$0-1] },{ t: 'terminator', l: $$[$0] }] } 
break;
case 6:
 this.$ = { nt: 'type_decrs', children: [$$[$0-1],$$[$0]] } 
break;
case 7:
 this.$ = { nt: 'type_decrs', children: [$$[$0]] } 
break;
case 8:
 this.$ = { nt: 'type_decr', children: [$$[$0]] } 
break;
case 9:
 this.$ = { nt: 'type_decr', children: [{ t: 'terminator', l: $$[$0] }] } 
break;
case 10:
 this.$ = { nt: 'class_decr', children: [{ t: 'public', l: $$[$0-4] },{ t: 'class', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 11:
 this.$ = { nt: 'class_decr', children: [{ t: 'class', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 12:
 this.$ = { nt: 'class_decr', children: [{ t: 'public', l: $$[$0-3] },{ t: 'class', l: $$[$0-2] },{ t: 'identifier', l: $$[$0-1] },$$[$0]] } 
break;
case 13:
 this.$ = { nt: 'class_decr', children: [{ t: 'class', l: $$[$0-2] },{ t: 'identifier', l: $$[$0-1] },$$[$0]] } 
break;
case 14:
 this.$ = { nt: 'extend_decr', children: [{ t: 'extends', l: $$[$0-1] },{ t: 'identifier', l: $$[$0] }] } 
break;
case 15:
 this.$ = { nt: 'class_body', children: [{ t: 'set_start', l: $$[$0-2] },$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 16:
 this.$ = { nt: 'class_body_decrs', children: [$$[$0-1],$$[$0]] } 
break;
case 17:
 this.$ = { nt: 'class_body_decrs', children: [$$[$0]] } 
break;
case 18:
 this.$ = { nt: 'class_body_decr', children: [$$[$0]] } 
break;
case 19:
 this.$ = { nt: 'class_body_decr', children: [{ t: 'public', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 20:
 this.$ = { nt: 'class_body_decr', children: [$$[$0-1],$$[$0]] } 
break;
case 21: case 22:
 this.$ = { nt: 'class_member_decr', children: [$$[$0]] } 
break;
case 23:
 this.$ = { nt: 'consr_declarator', children: [{ t: 'identifier', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 24:
 this.$ = { nt: 'formal_parameter_list', children: [$$[$0-2],{ t: 'separator', l: $$[$0-1] },$$[$0]] } 
break;
case 25:
 this.$ = { nt: 'formal_parameter_list', children: [$$[$0]] } 
break;
case 26:
 this.$ = { nt: 'formal_parameter_list', children: [] } 
break;
case 27:
 this.$ = { nt: 'formal_parameter', children: [$$[$0-1],$$[$0]] } 
break;
case 28:
 this.$ = { nt: 'consr_body', children: [{ t: 'set_start', l: $$[$0-3] },$$[$0-2],$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 29: case 30:
 this.$ = { nt: 'consr_body', children: [{ t: 'set_start', l: $$[$0-2] },$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 31:
 this.$ = { nt: 'consr_body', children: [{ t: 'set_start', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 32:
 this.$ = { nt: 'explicit_consr_invocation', children: [{ t: 'this', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 33:
 this.$ = { nt: 'explicit_consr_invocation', children: [{ t: 'super', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 34:
 this.$ = { nt: 'explicit_consr_invocation', children: [{ t: 'this', l: $$[$0-2] },{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 35:
 this.$ = { nt: 'explicit_consr_invocation', children: [{ t: 'super', l: $$[$0-2] },{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 36:
 this.$ = { nt: 'field_decr', children: [{ t: 'public', l: $$[$0-3] },$$[$0-2],$$[$0-1],{ t: 'terminator', l: $$[$0] }] } 
break;
case 37:
 this.$ = { nt: 'field_decr', children: [$$[$0-2],$$[$0-1],{ t: 'terminator', l: $$[$0] }] } 
break;
case 38:
 this.$ = { nt: 'var_declarators', children: [$$[$0-2],{ t: 'separator', l: $$[$0-1] },$$[$0]] } 
break;
case 39:
 this.$ = { nt: 'var_declarators', children: [$$[$0]] } 
break;
case 40:
 this.$ = { nt: 'var_declarator', children: [$$[$0]] } 
break;
case 41:
 this.$ = { nt: 'var_declarator', children: [$$[$0-2],{ t: 'op_assign', l: $$[$0-1] },$$[$0]] } 
break;
case 42:
 this.$ = { nt: 'var_declarator_id', children: [$$[$0-2],{ t: 'brackets_start', l: $$[$0-1] },{ t: 'brackets_end', l: $$[$0] }] } 
break;
case 43:
 this.$ = { nt: 'var_declarator_id', children: [{ t: 'identifier', l: $$[$0] }] } 
break;
case 44: case 45:
 this.$ = { nt: 'var_init', children: [$$[$0]] } 
break;
case 46:
 this.$ = { nt: 'method_decr', children: [{ t: 'public', l: $$[$0-3] },{ t: 'void', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 47:
 this.$ = { nt: 'method_decr', children: [{ t: 'public', l: $$[$0-3] },$$[$0-2],$$[$0-1],$$[$0]] } 
break;
case 48:
 this.$ = { nt: 'method_decr', children: [{ t: 'void', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 49:
 this.$ = { nt: 'method_decr', children: [$$[$0-2],$$[$0-1],$$[$0]] } 
break;
case 50:
 this.$ = { nt: 'method_declarator', children: [{ t: 'identifier', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 51:
 this.$ = { nt: 'method_body', children: [$$[$0]] } 
break;
case 52:
 this.$ = { nt: 'array_init', children: [{ t: 'set_start', l: $$[$0-3] },$$[$0-2],{ t: 'separator', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 53:
 this.$ = { nt: 'array_init', children: [{ t: 'set_start', l: $$[$0-2] },$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 54:
 this.$ = { nt: 'array_init', children: [{ t: 'set_start', l: $$[$0-2] },{ t: 'separator', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 55:
 this.$ = { nt: 'array_init', children: [{ t: 'set_start', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 56:
 this.$ = { nt: 'var_inits', children: [$$[$0-2],{ t: 'separator', l: $$[$0-1] },$$[$0]] } 
break;
case 57:
 this.$ = { nt: 'var_inits', children: [$$[$0]] } 
break;
case 58: case 59:
 this.$ = { nt: 'type', children: [$$[$0]] } 
break;
case 60: case 61:
 this.$ = { nt: 'primitive_type', children: [$$[$0]] } 
break;
case 62:
 this.$ = { nt: 'primitive_type', children: [{ t: 'boolean', l: $$[$0] }] } 
break;
case 63:
 this.$ = { nt: 'integral_type', children: [{ t: 'byte', l: $$[$0] }] } 
break;
case 64:
 this.$ = { nt: 'integral_type', children: [{ t: 'short', l: $$[$0] }] } 
break;
case 65:
 this.$ = { nt: 'integral_type', children: [{ t: 'int', l: $$[$0] }] } 
break;
case 66:
 this.$ = { nt: 'integral_type', children: [{ t: 'long', l: $$[$0] }] } 
break;
case 67:
 this.$ = { nt: 'integral_type', children: [{ t: 'char', l: $$[$0] }] } 
break;
case 68:
 this.$ = { nt: 'floating_type', children: [{ t: 'float', l: $$[$0] }] } 
break;
case 69:
 this.$ = { nt: 'floating_type', children: [{ t: 'double', l: $$[$0] }] } 
break;
case 70:
 this.$ = { nt: 'reference_type', children: [{ t: 'identifier', l: $$[$0] }] } 
break;
case 71:
 this.$ = { nt: 'reference_type', children: [$$[$0-2],{ t: 'brackets_start', l: $$[$0-1] },{ t: 'brackets_end', l: $$[$0] }] } 
break;
case 72:
 this.$ = { nt: 'block', children: [{ t: 'set_start', l: $$[$0-2] },$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 73:
 this.$ = { nt: 'block', children: [{ t: 'set_start', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 74:
 this.$ = { nt: 'block_stmts', children: [$$[$0-1],$$[$0]] } 
break;
case 75:
 this.$ = { nt: 'block_stmts', children: [$$[$0]] } 
break;
case 76:
 this.$ = { nt: 'block_stmt', children: [$$[$0-2],$$[$0-1],{ t: 'terminator', l: $$[$0] }] } 
break;
case 77:
 this.$ = { nt: 'block_stmt', children: [$$[$0]] } 
break;
case 78: case 79: case 80: case 81: case 82:
 this.$ = { nt: 'stmt', children: [$$[$0]] } 
break;
case 83: case 84: case 85: case 86:
 this.$ = { nt: 'stmt_nsi', children: [$$[$0]] } 
break;
case 87: case 88: case 89: case 90: case 91: case 92:
 this.$ = { nt: 'stmt_wots', children: [$$[$0]] } 
break;
case 93:
 this.$ = { nt: 'stmt_wots', children: [$$[$0-1],{ t: 'terminator', l: $$[$0] }] } 
break;
case 94:
 this.$ = { nt: 'stmt_wots', children: [{ t: 'terminator', l: $$[$0] }] } 
break;
case 95:
 this.$ = { nt: 'if_then_stmt', children: [{ t: 'if', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 96:
 this.$ = { nt: 'if_then_else_stmt', children: [{ t: 'if', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },$$[$0-4],{ t: 'paranthesis_end', l: $$[$0-3] },$$[$0-2],{ t: 'else', l: $$[$0-1] },$$[$0]] } 
break;
case 97:
 this.$ = { nt: 'if_then_else_stmt_nsi', children: [{ t: 'if', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },$$[$0-4],{ t: 'paranthesis_end', l: $$[$0-3] },$$[$0-2],{ t: 'else', l: $$[$0-1] },$$[$0]] } 
break;
case 98:
 this.$ = { nt: 'switch_stmt', children: [{ t: 'switch', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 99:
 this.$ = { nt: 'switch_block', children: [{ t: 'set_start', l: $$[$0-3] },$$[$0-2],$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 100: case 101:
 this.$ = { nt: 'switch_block', children: [{ t: 'set_start', l: $$[$0-2] },$$[$0-1],{ t: 'set_end', l: $$[$0] }] } 
break;
case 102:
 this.$ = { nt: 'switch_block', children: [{ t: 'set_start', l: $$[$0-1] },{ t: 'set_end', l: $$[$0] }] } 
break;
case 103:
 this.$ = { nt: 'switch_block_stmt_groups', children: [$$[$0-1],$$[$0]] } 
break;
case 104:
 this.$ = { nt: 'switch_block_stmt_groups', children: [$$[$0]] } 
break;
case 105:
 this.$ = { nt: 'switch_block_stmt_group', children: [$$[$0-1],$$[$0]] } 
break;
case 106:
 this.$ = { nt: 'switch_labels', children: [$$[$0-1],$$[$0]] } 
break;
case 107:
 this.$ = { nt: 'switch_labels', children: [$$[$0]] } 
break;
case 108:
 this.$ = { nt: 'switch_label', children: [{ t: 'case', l: $$[$0-2] },$$[$0-1],{ t: 'colon', l: $$[$0] }] } 
break;
case 109:
 this.$ = { nt: 'switch_label', children: [{ t: 'case', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },{ t: 'colon', l: $$[$0] }] } 
break;
case 110:
 this.$ = { nt: 'switch_label', children: [{ t: 'default', l: $$[$0-1] },{ t: 'colon', l: $$[$0] }] } 
break;
case 111:
 this.$ = { nt: 'while_stmt', children: [{ t: 'while', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 112:
 this.$ = { nt: 'while_stmt_nsi', children: [{ t: 'while', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 113:
 this.$ = { nt: 'do_stmt', children: [{ t: 'do', l: $$[$0-6] },$$[$0-5],{ t: 'while', l: $$[$0-4] },{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },{ t: 'terminator', l: $$[$0] }] } 
break;
case 114:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-8] },{ t: 'paranthesis_start', l: $$[$0-7] },$$[$0-6],{ t: 'terminator', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-5] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 115:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },{ t: 'terminator', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-5] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 116:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },$$[$0-5],{ t: 'terminator', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-4] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 117:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },$$[$0-5],{ t: 'terminator', l: $$[$0-4] },$$[$0-3],{ t: 'terminator', l: $$[$0-4] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 118:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },{ t: 'terminator', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-4] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 119:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },{ t: 'terminator', l: $$[$0-4] },$$[$0-3],{ t: 'terminator', l: $$[$0-4] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 120:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-3] },{ t: 'terminator', l: $$[$0-3] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 121:
 this.$ = { nt: 'for_stmt', children: [{ t: 'for', l: $$[$0-5] },{ t: 'paranthesis_start', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-3] },{ t: 'terminator', l: $$[$0-3] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 122:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-8] },{ t: 'paranthesis_start', l: $$[$0-7] },$$[$0-6],{ t: 'terminator', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-5] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 123:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },{ t: 'terminator', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-5] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 124:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },$$[$0-5],{ t: 'terminator', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-4] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 125:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-7] },{ t: 'paranthesis_start', l: $$[$0-6] },$$[$0-5],{ t: 'terminator', l: $$[$0-4] },$$[$0-3],{ t: 'terminator', l: $$[$0-4] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 126:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },{ t: 'terminator', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-4] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 127:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },{ t: 'terminator', l: $$[$0-4] },$$[$0-3],{ t: 'terminator', l: $$[$0-4] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 128:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-6] },{ t: 'paranthesis_start', l: $$[$0-5] },$$[$0-4],{ t: 'terminator', l: $$[$0-3] },{ t: 'terminator', l: $$[$0-3] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 129:
 this.$ = { nt: 'for_stmt_nsi', children: [{ t: 'for', l: $$[$0-5] },{ t: 'paranthesis_start', l: $$[$0-4] },{ t: 'terminator', l: $$[$0-3] },{ t: 'terminator', l: $$[$0-3] },{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 130:
 this.$ = { nt: 'for_init', children: [$$[$0]] } 
break;
case 131:
 this.$ = { nt: 'for_init', children: [$$[$0-1],$$[$0]] } 
break;
case 132:
 this.$ = { nt: 'stmt_expr_list', children: [$$[$0-2],{ t: 'separator', l: $$[$0-1] },$$[$0]] } 
break;
case 133:
 this.$ = { nt: 'stmt_expr_list', children: [$$[$0]] } 
break;
case 134:
 this.$ = { nt: 'break_stmt', children: [{ t: 'break', l: $$[$0-1] },{ t: 'terminator', l: $$[$0] }] } 
break;
case 135:
 this.$ = { nt: 'continue_stmt', children: [{ t: 'continue', l: $$[$0-1] },{ t: 'terminator', l: $$[$0] }] } 
break;
case 136:
 this.$ = { nt: 'return_stmt', children: [{ t: 'return', l: $$[$0-2] },$$[$0-1],{ t: 'terminator', l: $$[$0] }] } 
break;
case 137:
 this.$ = { nt: 'return_stmt', children: [{ t: 'return', l: $$[$0-1] },{ t: 'terminator', l: $$[$0] }] } 
break;
case 138: case 139:
 this.$ = { nt: 'expr', children: [$$[$0]] } 
break;
case 140: case 141: case 142: case 143: case 144: case 145:
 this.$ = { nt: 'stmt_expr', children: [$$[$0]] } 
break;
case 146:
 this.$ = { nt: 'assignment', children: [$$[$0-2],$$[$0-1],$$[$0]] } 
break;
case 147: case 148: case 149:
 this.$ = { nt: 'left_hand_side', children: [$$[$0]] } 
break;
case 150:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_assign', l: $$[$0] }] } 
break;
case 151:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_mulAssign', l: $$[$0] }] } 
break;
case 152:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_divAssign', l: $$[$0] }] } 
break;
case 153:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_modAssign', l: $$[$0] }] } 
break;
case 154:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_addAssign', l: $$[$0] }] } 
break;
case 155:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_subAssign', l: $$[$0] }] } 
break;
case 156:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_LshiftEqual', l: $$[$0] }] } 
break;
case 157:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_RshiftEqual', l: $$[$0] }] } 
break;
case 158:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_andAssign', l: $$[$0] }] } 
break;
case 159:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_orAssign', l: $$[$0] }] } 
break;
case 160:
 this.$ = { nt: 'assignment_operator', children: [{ t: 'op_xorAssign', l: $$[$0] }] } 
break;
case 161:
 this.$ = { nt: 'cond_expr', children: [$$[$0]] } 
break;
case 162:
 this.$ = { nt: 'cond_or_expr', children: [$$[$0]] } 
break;
case 163:
 this.$ = { nt: 'cond_or_expr', children: [$$[$0-2],{ t: 'op_oror', l: $$[$0-1] },$$[$0]] } 
break;
case 164:
 this.$ = { nt: 'cond_and_expr', children: [$$[$0]] } 
break;
case 165:
 this.$ = { nt: 'cond_and_expr', children: [$$[$0-2],{ t: 'op_andand', l: $$[$0-1] },$$[$0]] } 
break;
case 166:
 this.$ = { nt: 'incl_or_expr', children: [$$[$0]] } 
break;
case 167:
 this.$ = { nt: 'incl_or_expr', children: [$$[$0-2],{ t: 'op_or', l: $$[$0-1] },$$[$0]] } 
break;
case 168:
 this.$ = { nt: 'excl_or_expr', children: [$$[$0]] } 
break;
case 169:
 this.$ = { nt: 'excl_or_expr', children: [$$[$0-2],{ t: 'op_xor', l: $$[$0-1] },$$[$0]] } 
break;
case 170:
 this.$ = { nt: 'and_expr', children: [$$[$0]] } 
break;
case 171:
 this.$ = { nt: 'and_expr', children: [$$[$0-2],{ t: 'op_and', l: $$[$0-1] },$$[$0]] } 
break;
case 172:
 this.$ = { nt: 'equality_expr', children: [$$[$0]] } 
break;
case 173:
 this.$ = { nt: 'equality_expr', children: [$$[$0-2],{ t: 'op_equalCompare', l: $$[$0-1] },$$[$0]] } 
break;
case 174:
 this.$ = { nt: 'equality_expr', children: [$$[$0-2],{ t: 'op_notequalCompare', l: $$[$0-1] },$$[$0]] } 
break;
case 175:
 this.$ = { nt: 'relational_expr', children: [$$[$0]] } 
break;
case 176:
 this.$ = { nt: 'relational_expr', children: [$$[$0-2],{ t: 'op_greater', l: $$[$0-1] },$$[$0]] } 
break;
case 177:
 this.$ = { nt: 'relational_expr', children: [$$[$0-2],{ t: 'op_greaterEqual', l: $$[$0-1] },$$[$0]] } 
break;
case 178:
 this.$ = { nt: 'relational_expr', children: [$$[$0-2],{ t: 'op_less', l: $$[$0-1] },$$[$0]] } 
break;
case 179:
 this.$ = { nt: 'relational_expr', children: [$$[$0-2],{ t: 'op_lessEqual', l: $$[$0-1] },$$[$0]] } 
break;
case 180:
 this.$ = { nt: 'relational_expr', children: [$$[$0-2],{ t: 'instanceof', l: $$[$0-1] },$$[$0]] } 
break;
case 181:
 this.$ = { nt: 'shift_expr', children: [$$[$0]] } 
break;
case 182:
 this.$ = { nt: 'shift_expr', children: [$$[$0-2],{ t: 'op_Lshift', l: $$[$0-1] },$$[$0]] } 
break;
case 183:
 this.$ = { nt: 'shift_expr', children: [$$[$0-2],{ t: 'op_Rshift', l: $$[$0-1] },$$[$0]] } 
break;
case 184:
 this.$ = { nt: 'additive_expr', children: [$$[$0]] } 
break;
case 185:
 this.$ = { nt: 'additive_expr', children: [$$[$0-2],{ t: 'op_add', l: $$[$0-1] },$$[$0]] } 
break;
case 186:
 this.$ = { nt: 'additive_expr', children: [$$[$0-2],{ t: 'op_sub', l: $$[$0-1] },$$[$0]] } 
break;
case 187:
 this.$ = { nt: 'multiplicative_expr', children: [$$[$0]] } 
break;
case 188:
 this.$ = { nt: 'multiplicative_expr', children: [$$[$0-2],{ t: 'op_mul', l: $$[$0-1] },$$[$0]] } 
break;
case 189:
 this.$ = { nt: 'multiplicative_expr', children: [$$[$0-2],{ t: 'op_div', l: $$[$0-1] },$$[$0]] } 
break;
case 190:
 this.$ = { nt: 'multiplicative_expr', children: [$$[$0-2],{ t: 'op_mod', l: $$[$0-1] },$$[$0]] } 
break;
case 191:
 this.$ = { nt: 'predec_expr', children: [{ t: 'op_decrement', l: $$[$0-1] },$$[$0]] } 
break;
case 192:
 this.$ = { nt: 'preinc_expr', children: [{ t: 'op_increment', l: $$[$0-1] },$$[$0]] } 
break;
case 193: case 194: case 196:
 this.$ = { nt: 'unary_expr', children: [$$[$0]] } 
break;
case 195:
 this.$ = { nt: 'unary_expr', children: [$$[$0-1],$$[$0]] } 
break;
case 197: case 198: case 200:
 this.$ = { nt: 'unary_expr_npm', children: [$$[$0]] } 
break;
case 199:
 this.$ = { nt: 'unary_expr_npm', children: [{ t: 'op_not', l: $$[$0-1] },$$[$0]] } 
break;
case 201:
 this.$ = { nt: 'cast_expr', children: [{ t: 'paranthesis_start', l: $$[$0-3] },$$[$0-2],{ t: 'paranthesis_end', l: $$[$0-1] },$$[$0]] } 
break;
case 202: case 203:
 this.$ = { nt: 'postdec_expr', children: [$$[$0-1],{ t: 'op_decrement', l: $$[$0] }] } 
break;
case 204: case 205:
 this.$ = { nt: 'postinc_expr', children: [$$[$0-1],{ t: 'op_increment', l: $$[$0] }] } 
break;
case 206: case 207:
 this.$ = { nt: 'post_expr', children: [$$[$0]] } 
break;
case 208: case 209:
 this.$ = { nt: 'postfix_expr', children: [$$[$0]] } 
break;
case 210:
 this.$ = { nt: 'method_invocation', children: [$$[$0-3],{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 211:
 this.$ = { nt: 'method_invocation', children: [$$[$0-5],{ t: 'field_invoker', l: $$[$0-4] },{ t: 'identifier', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 212:
 this.$ = { nt: 'method_invocation', children: [{ t: 'super', l: $$[$0-5] },{ t: 'field_invoker', l: $$[$0-4] },{ t: 'identifier', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 213:
 this.$ = { nt: 'method_invocation', children: [$$[$0-2],{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 214:
 this.$ = { nt: 'method_invocation', children: [$$[$0-4],{ t: 'field_invoker', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 215:
 this.$ = { nt: 'method_invocation', children: [{ t: 'super', l: $$[$0-4] },{ t: 'field_invoker', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 216:
 this.$ = { nt: 'field_access', children: [$$[$0-2],{ t: 'field_invoker', l: $$[$0-1] },{ t: 'identifier', l: $$[$0] }] } 
break;
case 217:
 this.$ = { nt: 'field_access', children: [{ t: 'super', l: $$[$0-2] },{ t: 'field_invoker', l: $$[$0-1] },{ t: 'identifier', l: $$[$0] }] } 
break;
case 218: case 219:
 this.$ = { nt: 'array_access', children: [$$[$0-2],{ t: 'colon', l: $$[$0-1] },$$[$0]] } 
break;
case 220: case 221:
 this.$ = { nt: 'primary', children: [$$[$0]] } 
break;
case 222: case 225: case 226: case 227: case 228:
 this.$ = { nt: 'primary_no_new_array', children: [$$[$0]] } 
break;
case 223:
 this.$ = { nt: 'primary_no_new_array', children: [{ t: 'this', l: $$[$0] }] } 
break;
case 224:
 this.$ = { nt: 'primary_no_new_array', children: [{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 229:
 this.$ = { nt: 'class_instance_creation_expr', children: [{ t: 'new', l: $$[$0-4] },{ t: 'identifier', l: $$[$0-3] },{ t: 'paranthesis_start', l: $$[$0-2] },$$[$0-1],{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 230:
 this.$ = { nt: 'class_instance_creation_expr', children: [{ t: 'new', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },{ t: 'paranthesis_start', l: $$[$0-1] },{ t: 'paranthesis_end', l: $$[$0] }] } 
break;
case 231:
 this.$ = { nt: 'argument_list', children: [$$[$0]] } 
break;
case 232:
 this.$ = { nt: 'argument_list', children: [$$[$0-2],{ t: 'separator', l: $$[$0-1] },$$[$0]] } 
break;
case 233:
 this.$ = { nt: 'array_creation_expr', children: [{ t: 'new', l: $$[$0-3] },$$[$0-2],$$[$0-1],$$[$0]] } 
break;
case 234:
 this.$ = { nt: 'array_creation_expr', children: [{ t: 'new', l: $$[$0-3] },{ t: 'identifier', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 235:
 this.$ = { nt: 'array_creation_expr', children: [{ t: 'new', l: $$[$0-2] },$$[$0-1],$$[$0]] } 
break;
case 236:
 this.$ = { nt: 'array_creation_expr', children: [{ t: 'new', l: $$[$0-2] },{ t: 'identifier', l: $$[$0-1] },$$[$0]] } 
break;
case 237:
 this.$ = { nt: 'dim_exprs', children: [$$[$0-1],$$[$0]] } 
break;
case 238:
 this.$ = { nt: 'dim_exprs', children: [$$[$0]] } 
break;
case 239:
 this.$ = { nt: 'dim_expr', children: [{ t: 'brackets_start', l: $$[$0-2] },$$[$0-1],{ t: 'brackets_end', l: $$[$0] }] } 
break;
case 240:
 this.$ = { nt: 'dims', children: [{ t: 'brackets_start', l: $$[$0-1] },{ t: 'brackets_end', l: $$[$0] }] } 
break;
case 241:
 this.$ = { nt: 'dims', children: [$$[$0-2],{ t: 'brackets_start', l: $$[$0-1] },{ t: 'brackets_end', l: $$[$0] }] } 
break;
case 242:
 this.$ = { nt: 'expr_name', children: [{ t: 'identifier', l: $$[$0] }] } 
break;
case 243:
 this.$ = { nt: 'expr_name', children: [$$[$0-2],{ t: 'field_invoker', l: $$[$0-1] },{ t: 'identifier', l: $$[$0] }] } 
break;
case 244:
 this.$ = { nt: 'literal', children: [{ t: 'integer_literal', l: $$[$0] }] } 
break;
case 245:
 this.$ = { nt: 'literal', children: [{ t: 'float_literal', l: $$[$0] }] } 
break;
case 246:
 this.$ = { nt: 'literal', children: [{ t: 'boolean_literal', l: $$[$0] }] } 
break;
case 247:
 this.$ = { nt: 'literal', children: [{ t: 'character_literal', l: $$[$0] }] } 
break;
case 248:
 this.$ = { nt: 'literal', children: [{ t: 'string_literal', l: $$[$0] }] } 
break;
case 249:
 this.$ = { nt: 'literal', children: [{ t: 'null_literal', l: $$[$0] }] } 
break;
case 250:
 this.$ = { nt: 'sign', children: [{ t: 'op_add', l: $$[$0] }] } 
break;
case 251:
 this.$ = { nt: 'sign', children: [{ t: 'op_sub', l: $$[$0] }] } 
break;
}
},
table: [{3:1,4:2,5:[1,3],6:4,7:$V0},{1:[3]},{5:[1,6],6:7,7:$V0},{1:[2,2]},o($V1,[2,3]),{8:[1,8]},{1:[2,1]},o($V1,[2,4]),{9:[1,9]},o($V1,[2,5])],
defaultActions: {3:[2,2],6:[2,1]},
parseError: function parseError(str, hash) {
    if (hash.recoverable) {
        this.trace(str);
    } else {
        var error = new Error(str);
        error.hash = hash;
        throw error;
    }
},
parse: function parse(input) {
    var self = this, stack = [0], tstack = [], vstack = [null], lstack = [], table = this.table, yytext = '', yylineno = 0, yyleng = 0, recovering = 0, TERROR = 2, EOF = 1;
    var args = lstack.slice.call(arguments, 1);
    var lexer = Object.create(this.lexer);
    var sharedState = { yy: {} };
    for (var k in this.yy) {
        if (Object.prototype.hasOwnProperty.call(this.yy, k)) {
            sharedState.yy[k] = this.yy[k];
        }
    }
    lexer.setInput(input, sharedState.yy);
    sharedState.yy.lexer = lexer;
    sharedState.yy.parser = this;
    if (typeof lexer.yylloc == 'undefined') {
        lexer.yylloc = {};
    }
    var yyloc = lexer.yylloc;
    lstack.push(yyloc);
    var ranges = lexer.options && lexer.options.ranges;
    if (typeof sharedState.yy.parseError === 'function') {
        this.parseError = sharedState.yy.parseError;
    } else {
        this.parseError = Object.getPrototypeOf(this).parseError;
    }
    function popStack(n) {
        stack.length = stack.length - 2 * n;
        vstack.length = vstack.length - n;
        lstack.length = lstack.length - n;
    }
    _token_stack:
        var lex = function () {
            var token;
            token = lexer.lex() || EOF;
            if (typeof token !== 'number') {
                token = self.symbols_[token] || token;
            }
            return token;
        };
    var symbol, preErrorSymbol, state, action, a, r, yyval = {}, p, len, newState, expected;
    while (true) {
        state = stack[stack.length - 1];
        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol === null || typeof symbol == 'undefined') {
                symbol = lex();
            }
            action = table[state] && table[state][symbol];
        }
                    if (typeof action === 'undefined' || !action.length || !action[0]) {
                var errStr = '';
                expected = [];
                for (p in table[state]) {
                    if (this.terminals_[p] && p > TERROR) {
                        expected.push('\'' + this.terminals_[p] + '\'');
                    }
                }
                if (lexer.showPosition) {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ':\n' + lexer.showPosition() + '\nExpecting ' + expected.join(', ') + ', got \'' + (this.terminals_[symbol] || symbol) + '\'';
                } else {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ': Unexpected ' + (symbol == EOF ? 'end of input' : '\'' + (this.terminals_[symbol] || symbol) + '\'');
                }
                this.parseError(errStr, {
                    text: lexer.match,
                    token: this.terminals_[symbol] || symbol,
                    line: lexer.yylineno,
                    loc: yyloc,
                    expected: expected
                });
            }
        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
        }
        switch (action[0]) {
        case 1:
            stack.push(symbol);
            vstack.push(lexer.yytext);
            lstack.push(lexer.yylloc);
            stack.push(action[1]);
            symbol = null;
            if (!preErrorSymbol) {
                yyleng = lexer.yyleng;
                yytext = lexer.yytext;
                yylineno = lexer.yylineno;
                yyloc = lexer.yylloc;
                if (recovering > 0) {
                    recovering--;
                }
            } else {
                symbol = preErrorSymbol;
                preErrorSymbol = null;
            }
            break;
        case 2:
            len = this.productions_[action[1]][1];
            yyval.$ = vstack[vstack.length - len];
            yyval._$ = {
                first_line: lstack[lstack.length - (len || 1)].first_line,
                last_line: lstack[lstack.length - 1].last_line,
                first_column: lstack[lstack.length - (len || 1)].first_column,
                last_column: lstack[lstack.length - 1].last_column
            };
            if (ranges) {
                yyval._$.range = [
                    lstack[lstack.length - (len || 1)].range[0],
                    lstack[lstack.length - 1].range[1]
                ];
            }
            r = this.performAction.apply(yyval, [
                yytext,
                yyleng,
                yylineno,
                sharedState.yy,
                action[1],
                vstack,
                lstack
            ].concat(args));
            if (typeof r !== 'undefined') {
                return r;
            }
            if (len) {
                stack = stack.slice(0, -1 * len * 2);
                vstack = vstack.slice(0, -1 * len);
                lstack = lstack.slice(0, -1 * len);
            }
            stack.push(this.productions_[action[1]][0]);
            vstack.push(yyval.$);
            lstack.push(yyval._$);
            newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
            stack.push(newState);
            break;
        case 3:
            return true;
        }
    }
    return true;
}};
/* generated by jison-lex 0.3.4 */
var lexer = (function(){
var lexer = ({

EOF:1,

parseError:function parseError(str, hash) {
        if (this.yy.parser) {
            this.yy.parser.parseError(str, hash);
        } else {
            throw new Error(str);
        }
    },

// resets the lexer, sets new input
setInput:function (input, yy) {
        this.yy = yy || this.yy || {};
        this._input = input;
        this._more = this._backtrack = this.done = false;
        this.yylineno = this.yyleng = 0;
        this.yytext = this.matched = this.match = '';
        this.conditionStack = ['INITIAL'];
        this.yylloc = {
            first_line: 1,
            first_column: 0,
            last_line: 1,
            last_column: 0
        };
        if (this.options.ranges) {
            this.yylloc.range = [0,0];
        }
        this.offset = 0;
        return this;
    },

// consumes and returns one char from the input
input:function () {
        var ch = this._input[0];
        this.yytext += ch;
        this.yyleng++;
        this.offset++;
        this.match += ch;
        this.matched += ch;
        var lines = ch.match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno++;
            this.yylloc.last_line++;
        } else {
            this.yylloc.last_column++;
        }
        if (this.options.ranges) {
            this.yylloc.range[1]++;
        }

        this._input = this._input.slice(1);
        return ch;
    },

// unshifts one char (or a string) into the input
unput:function (ch) {
        var len = ch.length;
        var lines = ch.split(/(?:\r\n?|\n)/g);

        this._input = ch + this._input;
        this.yytext = this.yytext.substr(0, this.yytext.length - len);
        //this.yyleng -= len;
        this.offset -= len;
        var oldLines = this.match.split(/(?:\r\n?|\n)/g);
        this.match = this.match.substr(0, this.match.length - 1);
        this.matched = this.matched.substr(0, this.matched.length - 1);

        if (lines.length - 1) {
            this.yylineno -= lines.length - 1;
        }
        var r = this.yylloc.range;

        this.yylloc = {
            first_line: this.yylloc.first_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.first_column,
            last_column: lines ?
                (lines.length === oldLines.length ? this.yylloc.first_column : 0)
                 + oldLines[oldLines.length - lines.length].length - lines[0].length :
              this.yylloc.first_column - len
        };

        if (this.options.ranges) {
            this.yylloc.range = [r[0], r[0] + this.yyleng - len];
        }
        this.yyleng = this.yytext.length;
        return this;
    },

// When called from action, caches matched text and appends it on next action
more:function () {
        this._more = true;
        return this;
    },

// When called from action, signals the lexer that this rule fails to match the input, so the next matching rule (regex) should be tested instead.
reject:function () {
        if (this.options.backtrack_lexer) {
            this._backtrack = true;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. You can only invoke reject() in the lexer when the lexer is of the backtracking persuasion (options.backtrack_lexer = true).\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });

        }
        return this;
    },

// retain first n characters of the match
less:function (n) {
        this.unput(this.match.slice(n));
    },

// displays already matched input, i.e. for error messages
pastInput:function () {
        var past = this.matched.substr(0, this.matched.length - this.match.length);
        return (past.length > 20 ? '...':'') + past.substr(-20).replace(/\n/g, "");
    },

// displays upcoming input, i.e. for error messages
upcomingInput:function () {
        var next = this.match;
        if (next.length < 20) {
            next += this._input.substr(0, 20-next.length);
        }
        return (next.substr(0,20) + (next.length > 20 ? '...' : '')).replace(/\n/g, "");
    },

// displays the character position where the lexing error occurred, i.e. for error messages
showPosition:function () {
        var pre = this.pastInput();
        var c = new Array(pre.length + 1).join("-");
        return pre + this.upcomingInput() + "\n" + c + "^";
    },

// test the lexed token: return FALSE when not a match, otherwise return token
test_match:function (match, indexed_rule) {
        var token,
            lines,
            backup;

        if (this.options.backtrack_lexer) {
            // save context
            backup = {
                yylineno: this.yylineno,
                yylloc: {
                    first_line: this.yylloc.first_line,
                    last_line: this.last_line,
                    first_column: this.yylloc.first_column,
                    last_column: this.yylloc.last_column
                },
                yytext: this.yytext,
                match: this.match,
                matches: this.matches,
                matched: this.matched,
                yyleng: this.yyleng,
                offset: this.offset,
                _more: this._more,
                _input: this._input,
                yy: this.yy,
                conditionStack: this.conditionStack.slice(0),
                done: this.done
            };
            if (this.options.ranges) {
                backup.yylloc.range = this.yylloc.range.slice(0);
            }
        }

        lines = match[0].match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno += lines.length;
        }
        this.yylloc = {
            first_line: this.yylloc.last_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.last_column,
            last_column: lines ?
                         lines[lines.length - 1].length - lines[lines.length - 1].match(/\r?\n?/)[0].length :
                         this.yylloc.last_column + match[0].length
        };
        this.yytext += match[0];
        this.match += match[0];
        this.matches = match;
        this.yyleng = this.yytext.length;
        if (this.options.ranges) {
            this.yylloc.range = [this.offset, this.offset += this.yyleng];
        }
        this._more = false;
        this._backtrack = false;
        this._input = this._input.slice(match[0].length);
        this.matched += match[0];
        token = this.performAction.call(this, this.yy, this, indexed_rule, this.conditionStack[this.conditionStack.length - 1]);
        if (this.done && this._input) {
            this.done = false;
        }
        if (token) {
            return token;
        } else if (this._backtrack) {
            // recover context
            for (var k in backup) {
                this[k] = backup[k];
            }
            return false; // rule action called reject() implying the next rule should be tested instead.
        }
        return false;
    },

// return next match in input
next:function () {
        if (this.done) {
            return this.EOF;
        }
        if (!this._input) {
            this.done = true;
        }

        var token,
            match,
            tempMatch,
            index;
        if (!this._more) {
            this.yytext = '';
            this.match = '';
        }
        var rules = this._currentRules();
        for (var i = 0; i < rules.length; i++) {
            tempMatch = this._input.match(this.rules[rules[i]]);
            if (tempMatch && (!match || tempMatch[0].length > match[0].length)) {
                match = tempMatch;
                index = i;
                if (this.options.backtrack_lexer) {
                    token = this.test_match(tempMatch, rules[i]);
                    if (token !== false) {
                        return token;
                    } else if (this._backtrack) {
                        match = false;
                        continue; // rule action called reject() implying a rule MISmatch.
                    } else {
                        // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
                        return false;
                    }
                } else if (!this.options.flex) {
                    break;
                }
            }
        }
        if (match) {
            token = this.test_match(match, rules[index]);
            if (token !== false) {
                return token;
            }
            // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
            return false;
        }
        if (this._input === "") {
            return this.EOF;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. Unrecognized text.\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });
        }
    },

// return next match that has a token
lex:function lex() {
        var r = this.next();
        if (r) {
            return r;
        } else {
            return this.lex();
        }
    },

// activates a new lexer condition state (pushes the new lexer condition state onto the condition stack)
begin:function begin(condition) {
        this.conditionStack.push(condition);
    },

// pop the previously active lexer condition state off the condition stack
popState:function popState() {
        var n = this.conditionStack.length - 1;
        if (n > 0) {
            return this.conditionStack.pop();
        } else {
            return this.conditionStack[0];
        }
    },

// produce the lexer rule set which is active for the currently active lexer condition state
_currentRules:function _currentRules() {
        if (this.conditionStack.length && this.conditionStack[this.conditionStack.length - 1]) {
            return this.conditions[this.conditionStack[this.conditionStack.length - 1]].rules;
        } else {
            return this.conditions["INITIAL"].rules;
        }
    },

// return the currently active lexer condition state; when an index argument is provided it produces the N-th previous condition state, if available
topState:function topState(n) {
        n = this.conditionStack.length - 1 - Math.abs(n || 0);
        if (n >= 0) {
            return this.conditionStack[n];
        } else {
            return "INITIAL";
        }
    },

// alias for begin(condition)
pushState:function pushState(condition) {
        this.begin(condition);
    },

// return the number of states currently on the stack
stateStackSize:function stateStackSize() {
        return this.conditionStack.length;
    },
options: {},
performAction: function anonymous(yy,yy_,$avoiding_name_collisions,YY_START) {
var YYSTATE=YY_START;
switch($avoiding_name_collisions) {
case 0:this.pushState('BLOCKCOMMENT');
break;
case 1:this.popState();
break;
case 2:/* SKIP BLOCKCOMMENTS */
break;
case 3:this.pushState('COMMENT');
break;
case 4:/* SKIP COMMENTS */
break;
case 5:this.popState();
break;
case 6:/* SKIP WHITESPACES */
break;
case 7:return 56;
break;
case 8:return 98;
break;
case 9:return 57;
break;
case 10:return 89;
break;
case 11:return 61;
break;
case 12:return 14;
break;
case 13:return 'const';
break;
case 14:return 99;
break;
case 15:return 92;
break;
case 16:return 94;
break;
case 17:return 63;
break;
case 18:return 82;
break;
case 19:return 17;
break;
case 20:return 62;
break;
case 21:return 95;
break;
case 22:return 81;
break;
case 23:return 7;
break;
case 24:return 142;
break;
case 25:return 59;
break;
case 26:return 60;
break;
case 27:return 167;
break;
case 28:return 13;
break;
case 29:return 100;
break;
case 30:return 58;
break;
case 31:return 'static';
break;
case 32:return 38;
break;
case 33:return 83;
break;
case 34:return 36;
break;
case 35:return 47;
break;
case 36:return 93;
break;
case 37:return 154;
break;
case 38:return 153;
break;
case 39:return 116;
break;
case 40:return 117;
break;
case 41:return 113;
break;
case 42:return 114;
break;
case 43:return 115;
break;
case 44:return 120;
break;
case 45:return 121;
break;
case 46:return 122;
break;
case 47:return 136;
break;
case 48:return 135;
break;
case 49:return 118;
break;
case 50:return 119;
break;
case 51:return 139;
break;
case 52:return 141;
break;
case 53:return 144;
break;
case 54:return 145;
break;
case 55:return 147;
break;
case 56:return 148;
break;
case 57:return 150;
break;
case 58:return 151;
break;
case 59:return 152;
break;
case 60:return 138;
break;
case 61:return 140;
break;
case 62:return 41;
break;
case 63:return 127;
break;
case 64:return 125;
break;
case 65:return 133;
break;
case 66:return 129;
break;
case 67:return 158;
break;
case 68:return 131;
break;
case 69:return 91;
break;
case 70:return 171;
break;
case 71:return 170;
break;
case 72:return 172;
break;
case 73:return 172;
break;
case 74:return 175;
break;
case 75:return 173;
break;
case 76:return 174;
break;
case 77:return 8;
break;
case 78:return 9;
break;
case 79:return 163;
break;
case 80:return 30;
break;
case 81:return 27;
break;
case 82:return 29;
break;
case 83:return 43;
break;
case 84:return 44;
break;
case 85:return 18;
break;
case 86:return 20;
break;
case 87:return 5;
break;
}
},
rules: [/^(?:\/\*)/,/^(?:\*\/)/,/^(?:(\n|\r|.))/,/^(?:\/\/)/,/^(?:(.))/,/^(?:(\n|\r))/,/^(?:\s+)/,/^(?:boolean\b)/,/^(?:break\b)/,/^(?:byte\b)/,/^(?:case\b)/,/^(?:char\b)/,/^(?:class\b)/,/^(?:const\b)/,/^(?:continue\b)/,/^(?:default\b)/,/^(?:do\b)/,/^(?:double\b)/,/^(?:else\b)/,/^(?:extends\b)/,/^(?:float\b)/,/^(?:for\b)/,/^(?:if\b)/,/^(?:import\b)/,/^(?:instanceof\b)/,/^(?:int\b)/,/^(?:long\b)/,/^(?:new\b)/,/^(?:public\b)/,/^(?:return\b)/,/^(?:short\b)/,/^(?:static\b)/,/^(?:super\b)/,/^(?:switch\b)/,/^(?:this\b)/,/^(?:void\b)/,/^(?:while\b)/,/^(?:[+][+])/,/^(?:[-][-])/,/^(?:[+][=])/,/^(?:[-][=])/,/^(?:[*][=])/,/^(?:[\/][=])/,/^(?:[%][=])/,/^(?:[&][=])/,/^(?:[|][=])/,/^(?:[\^][=])/,/^(?:[!][=])/,/^(?:[=][=])/,/^(?:[<][<][=])/,/^(?:[>][>][=])/,/^(?:[>][=])/,/^(?:[<][=])/,/^(?:[<][<])/,/^(?:[>][>])/,/^(?:[+])/,/^(?:[-])/,/^(?:[*])/,/^(?:[\/])/,/^(?:[%])/,/^(?:[>])/,/^(?:[<])/,/^(?:[=])/,/^(?:[&][&])/,/^(?:[|][|])/,/^(?:[&])/,/^(?:[|])/,/^(?:[!])/,/^(?:[\^])/,/^(?:[:])/,/^(?:[0-9]+\.[0-9]*)/,/^(?:[0-9]+)/,/^(?:true\b)/,/^(?:false\b)/,/^(?:null\b)/,/^(?:'(\\[^\n\r]|[^\\\'\n\r])')/,/^(?:"(\\[^\n\r]|[^\\\'\n\r])*")/,/^(?:([a-z]|[A-Z]|[$]|[_])(\w)*)/,/^(?:[;])/,/^(?:[.])/,/^(?:[,])/,/^(?:[(])/,/^(?:[)])/,/^(?:[\[])/,/^(?:[\]])/,/^(?:[{])/,/^(?:[}])/,/^(?:$)/],
conditions: {"COMMENT":{"rules":[0,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87],"inclusive":true},"BLOCKCOMMENT":{"rules":[0,1,2,3,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87],"inclusive":true},"INITIAL":{"rules":[0,3,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87],"inclusive":true}}
});
return lexer;
})();
parser.lexer = lexer;
function Parser () {
  this.yy = {};
}
Parser.prototype = parser;parser.Parser = Parser;
return new Parser;
})();


if (typeof require !== 'undefined' && typeof exports !== 'undefined') {
exports.parser = index;
exports.Parser = index.Parser;
exports.parse = function () { return index.parse.apply(index, arguments); };
exports.main = function commonjsMain(args) {
    if (!args[1]) {
        console.log('Usage: '+args[0]+' FILE');
        process.exit(1);
    }
    var source = require('fs').readFileSync(require('path').normalize(args[1]), "utf8");
    return exports.parser.parse(source);
};
if (typeof module !== 'undefined' && require.main === module) {
  exports.main(process.argv.slice(1));
}
}
