// genlexer/index.js
var MyLexer = (function (undefined) {
function CDFA_base(){
	this.ss=undefined;
	this.as=undefined;
	this.tt=undefined;
this.stt={};
}
CDFA_base.prototype.reset = function (state) {
	this.cs = state || 	this.ss;
this.bol=false;
};
CDFA_base.prototype.readSymbol = function (c) {
	this.cs = this.nextState(this.cs, c);
};
CDFA_base.prototype.isAccepting = function () {
	var acc = this.as.indexOf(this.cs)>=0;
if((this.stt[this.cs]===-1)&&!this.bol){
acc=false;}
return acc;};
CDFA_base.prototype.isInDeadState = function () {
	return this.cs === undefined || this.cs === 0;
};
CDFA_base.prototype.getCurrentToken = function(){
	var t= this.tt[this.cs];
var s=this.stt[this.cs];
if(s!==undefined){return this.bol?t:s;}
return t;};

function CDFA_DEFAULT(){
	this.ss=1;
	this.as=[1,2,3,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,30,32,33,34,35,36,37,38,39,40,42,43,44,45,46,47,48,49,63,65,66,67,68,69,96,97,107,112,126,140,143,145,156,157,158,164,169,175,176,179,180,181,186,190,192,194,198,200,202,203,204,205,207,208,210,213];
	this.tt=[null,73,73,55,null,35,53,63,64,33,31,62,32,61,34,30,60,49,50,48,58,65,66,56,58,58,67,54,68,null,59,null,40,51,38,36,37,29,70,69,39,null,46,45,41,44,47,58,58,58,null,null,null,null,null,null,null,null,null,null,null,null,null,52,null,69,42,43,58,58,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,58,57,null,null,null,null,null,null,null,null,null,9,null,null,null,null,15,null,null,null,null,null,null,null,null,null,null,null,null,null,57,null,null,null,null,null,null,null,null,null,null,null,null,null,14,null,null,18,null,20,null,null,null,null,null,null,null,null,null,null,2,3,4,null,null,null,null,null,11,null,null,null,null,19,null,null,null,null,null,26,27,null,null,1,5,6,null,null,null,null,13,null,null,null,22,null,24,null,28,null,null,null,10,null,16,null,21,23,25,0,null,8,12,null,7,null,null,17];
this.stt={};
}
CDFA_DEFAULT.prototype= new CDFA_base();
CDFA_DEFAULT.prototype.nextState = function(state, c){
    var next = 0;
    switch(state){
case 1:
if(("\t" <= c && c <= "\n")  || ("\r" === c ) || (" " === c ) || (" " === c )){
next = 2;
} else if(("!" === c )){
next = 3;
} else if(("\"" === c )){
next = 4;
} else if(("%" === c )){
next = 5;
} else if(("&" === c )){
next = 6;
} else if(("(" === c )){
next = 7;
} else if((")" === c )){
next = 8;
} else if(("*" === c )){
next = 9;
} else if(("+" === c )){
next = 10;
} else if(("," === c )){
next = 11;
} else if(("-" === c )){
next = 12;
} else if(("." === c )){
next = 13;
} else if(("/" === c )){
next = 14;
} else if(("0" <= c && c <= "9") ){
next = 15;
} else if((";" === c )){
next = 16;
} else if(("<" === c )){
next = 17;
} else if(("=" === c )){
next = 18;
} else if((">" === c )){
next = 19;
} else if(("A" <= c && c <= "Z")  || ("a" <= c && c <= "e")  || ("g" <= c && c <= "s")  || ("u" <= c && c <= "z") ){
next = 20;
} else if(("[" === c )){
next = 21;
} else if(("]" === c )){
next = 22;
} else if(("^" === c )){
next = 23;
} else if(("f" === c )){
next = 24;
} else if(("t" === c )){
next = 25;
} else if(("{" === c )){
next = 26;
} else if(("|" === c )){
next = 27;
} else if(("}" === c )){
next = 28;
}
break;
case 2:
if(("\t" <= c && c <= "\n")  || ("\r" === c ) || (" " === c ) || (" " === c )){
next = 2;
}
break;
case 4:
if((c < "\"" || "\"" < c)  && (c < "\\" || "\\" < c) ){
next = 4;
} else if(("\"" === c )){
next = 30;
} else if(("\\" === c )){
next = 31;
}
break;
case 5:
if(("=" === c )){
next = 32;
}
break;
case 6:
if(("&" === c )){
next = 33;
}
break;
case 9:
if(("=" === c )){
next = 34;
}
break;
case 10:
if(("=" === c )){
next = 35;
}
break;
case 12:
if(("=" === c )){
next = 36;
}
break;
case 13:
if(("0" <= c && c <= "9") ){
next = 37;
}
break;
case 14:
if(("*" === c )){
next = 38;
} else if(("/" === c )){
next = 39;
} else if(("=" === c )){
next = 40;
}
break;
case 15:
if(("." === c )){
next = 41;
} else if(("0" <= c && c <= "9") ){
next = 15;
}
break;
case 17:
if(("<" === c )){
next = 42;
} else if(("=" === c )){
next = 43;
}
break;
case 18:
if(("=" === c )){
next = 44;
}
break;
case 19:
if(("=" === c )){
next = 45;
} else if((">" === c )){
next = 46;
}
break;
case 20:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "z") ){
next = 20;
}
break;
case 24:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("b" <= c && c <= "z") ){
next = 20;
} else if(("a" === c )){
next = 48;
}
break;
case 25:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "q")  || ("s" <= c && c <= "z") ){
next = 20;
} else if(("r" === c )){
next = 49;
}
break;
case 26:
if(("b" === c )){
next = 50;
} else if(("c" === c )){
next = 51;
} else if(("d" === c )){
next = 52;
} else if(("e" === c )){
next = 53;
} else if(("f" === c )){
next = 54;
} else if(("i" === c )){
next = 55;
} else if(("l" === c )){
next = 56;
} else if(("n" === c )){
next = 57;
} else if(("r" === c )){
next = 58;
} else if(("s" === c )){
next = 59;
} else if(("t" === c )){
next = 60;
} else if(("v" === c )){
next = 61;
} else if(("w" === c )){
next = 62;
}
break;
case 27:
if(("|" === c )){
next = 63;
}
break;
case 31:
if((c < "\n" || "\n" < c)  && (c < "\r" || "\r" < c) ){
next = 4;
}
break;
case 37:
if(("0" <= c && c <= "9") ){
next = 37;
}
break;
case 39:
if((c < "\n" || "\n" < c)  && (c < "\r" || "\r" < c) ){
next = 39;
}
break;
case 41:
if(("0" <= c && c <= "9") ){
next = 37;
}
break;
case 42:
if(("=" === c )){
next = 66;
}
break;
case 46:
if(("=" === c )){
next = 67;
}
break;
case 48:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "k")  || ("m" <= c && c <= "z") ){
next = 20;
} else if(("l" === c )){
next = 68;
}
break;
case 49:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "t")  || ("v" <= c && c <= "z") ){
next = 20;
} else if(("u" === c )){
next = 69;
}
break;
case 50:
if(("o" === c )){
next = 70;
} else if(("r" === c )){
next = 71;
} else if(("y" === c )){
next = 72;
}
break;
case 51:
if(("a" === c )){
next = 73;
} else if(("h" === c )){
next = 74;
} else if(("l" === c )){
next = 75;
} else if(("o" === c )){
next = 76;
}
break;
case 52:
if(("e" === c )){
next = 77;
} else if(("o" === c )){
next = 78;
}
break;
case 53:
if(("l" === c )){
next = 79;
} else if(("x" === c )){
next = 80;
}
break;
case 54:
if(("l" === c )){
next = 81;
} else if(("o" === c )){
next = 82;
}
break;
case 55:
if(("f" === c )){
next = 83;
} else if(("m" === c )){
next = 84;
} else if(("n" === c )){
next = 85;
}
break;
case 56:
if(("o" === c )){
next = 86;
}
break;
case 57:
if(("e" === c )){
next = 87;
}
break;
case 58:
if(("e" === c )){
next = 88;
}
break;
case 59:
if(("h" === c )){
next = 89;
} else if(("t" === c )){
next = 90;
} else if(("u" === c )){
next = 91;
} else if(("w" === c )){
next = 92;
}
break;
case 60:
if(("h" === c )){
next = 93;
}
break;
case 61:
if(("o" === c )){
next = 94;
}
break;
case 62:
if(("h" === c )){
next = 95;
}
break;
case 68:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "r")  || ("t" <= c && c <= "z") ){
next = 20;
} else if(("s" === c )){
next = 69;
}
break;
case 69:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "d")  || ("f" <= c && c <= "z") ){
next = 20;
} else if(("e" === c )){
next = 97;
}
break;
case 70:
if(("o" === c )){
next = 98;
}
break;
case 71:
if(("e" === c )){
next = 99;
}
break;
case 72:
if(("t" === c )){
next = 100;
}
break;
case 73:
if(("s" === c )){
next = 101;
}
break;
case 74:
if(("a" === c )){
next = 102;
}
break;
case 75:
if(("a" === c )){
next = 103;
}
break;
case 76:
if(("n" === c )){
next = 104;
}
break;
case 77:
if(("f" === c )){
next = 105;
}
break;
case 78:
if(("u" === c )){
next = 106;
} else if(("}" === c )){
next = 107;
}
break;
case 79:
if(("s" === c )){
next = 108;
}
break;
case 80:
if(("t" === c )){
next = 109;
}
break;
case 81:
if(("o" === c )){
next = 110;
}
break;
case 82:
if(("r" === c )){
next = 111;
}
break;
case 83:
if(("}" === c )){
next = 112;
}
break;
case 84:
if(("p" === c )){
next = 113;
}
break;
case 85:
if(("s" === c )){
next = 114;
} else if(("t" === c )){
next = 115;
}
break;
case 86:
if(("n" === c )){
next = 116;
}
break;
case 87:
if(("w" === c )){
next = 117;
}
break;
case 88:
if(("t" === c )){
next = 118;
}
break;
case 89:
if(("o" === c )){
next = 119;
}
break;
case 90:
if(("a" === c )){
next = 120;
}
break;
case 91:
if(("p" === c )){
next = 121;
}
break;
case 92:
if(("i" === c )){
next = 122;
}
break;
case 93:
if(("i" === c )){
next = 123;
}
break;
case 94:
if(("i" === c )){
next = 124;
}
break;
case 95:
if(("i" === c )){
next = 125;
}
break;
case 97:
if(("0" <= c && c <= "9")  || ("A" <= c && c <= "Z")  || ("_" === c ) || ("a" <= c && c <= "z") ){
next = 20;
}
break;
case 98:
if(("l" === c )){
next = 127;
}
break;
case 99:
if(("a" === c )){
next = 128;
}
break;
case 100:
if(("e" === c )){
next = 129;
}
break;
case 101:
if(("e" === c )){
next = 130;
}
break;
case 102:
if(("r" === c )){
next = 131;
}
break;
case 103:
if(("s" === c )){
next = 132;
}
break;
case 104:
if(("s" === c )){
next = 133;
} else if(("t" === c )){
next = 134;
}
break;
case 105:
if(("a" === c )){
next = 135;
}
break;
case 106:
if(("b" === c )){
next = 136;
}
break;
case 108:
if(("e" === c )){
next = 137;
}
break;
case 109:
if(("e" === c )){
next = 138;
}
break;
case 110:
if(("a" === c )){
next = 139;
}
break;
case 111:
if(("}" === c )){
next = 140;
}
break;
case 113:
if(("o" === c )){
next = 141;
}
break;
case 114:
if(("t" === c )){
next = 142;
}
break;
case 115:
if(("}" === c )){
next = 143;
}
break;
case 116:
if(("g" === c )){
next = 144;
}
break;
case 117:
if(("}" === c )){
next = 145;
}
break;
case 118:
if(("u" === c )){
next = 146;
}
break;
case 119:
if(("r" === c )){
next = 147;
}
break;
case 120:
if(("t" === c )){
next = 148;
}
break;
case 121:
if(("e" === c )){
next = 149;
}
break;
case 122:
if(("t" === c )){
next = 150;
}
break;
case 123:
if(("s" === c )){
next = 151;
}
break;
case 124:
if(("d" === c )){
next = 152;
}
break;
case 125:
if(("l" === c )){
next = 153;
}
break;
case 127:
if(("e" === c )){
next = 154;
}
break;
case 128:
if(("k" === c )){
next = 155;
}
break;
case 129:
if(("}" === c )){
next = 156;
}
break;
case 130:
if(("}" === c )){
next = 157;
}
break;
case 131:
if(("}" === c )){
next = 158;
}
break;
case 132:
if(("s" === c )){
next = 159;
}
break;
case 133:
if(("t" === c )){
next = 160;
}
break;
case 134:
if(("i" === c )){
next = 161;
}
break;
case 135:
if(("u" === c )){
next = 162;
}
break;
case 136:
if(("l" === c )){
next = 163;
}
break;
case 137:
if(("}" === c )){
next = 164;
}
break;
case 138:
if(("n" === c )){
next = 165;
}
break;
case 139:
if(("t" === c )){
next = 166;
}
break;
case 141:
if(("r" === c )){
next = 167;
}
break;
case 142:
if(("a" === c )){
next = 168;
}
break;
case 144:
if(("}" === c )){
next = 169;
}
break;
case 146:
if(("r" === c )){
next = 170;
}
break;
case 147:
if(("t" === c )){
next = 171;
}
break;
case 148:
if(("i" === c )){
next = 172;
}
break;
case 149:
if(("r" === c )){
next = 173;
}
break;
case 150:
if(("c" === c )){
next = 174;
}
break;
case 151:
if(("}" === c )){
next = 175;
}
break;
case 152:
if(("}" === c )){
next = 176;
}
break;
case 153:
if(("e" === c )){
next = 177;
}
break;
case 154:
if(("a" === c )){
next = 178;
}
break;
case 155:
if(("}" === c )){
next = 179;
}
break;
case 159:
if(("}" === c )){
next = 180;
}
break;
case 160:
if(("}" === c )){
next = 181;
}
break;
case 161:
if(("n" === c )){
next = 182;
}
break;
case 162:
if(("l" === c )){
next = 183;
}
break;
case 163:
if(("e" === c )){
next = 184;
}
break;
case 165:
if(("d" === c )){
next = 185;
}
break;
case 166:
if(("}" === c )){
next = 186;
}
break;
case 167:
if(("t" === c )){
next = 187;
}
break;
case 168:
if(("n" === c )){
next = 188;
}
break;
case 170:
if(("n" === c )){
next = 189;
}
break;
case 171:
if(("}" === c )){
next = 190;
}
break;
case 172:
if(("c" === c )){
next = 191;
}
break;
case 173:
if(("}" === c )){
next = 192;
}
break;
case 174:
if(("h" === c )){
next = 193;
}
break;
case 177:
if(("}" === c )){
next = 194;
}
break;
case 178:
if(("n" === c )){
next = 195;
}
break;
case 182:
if(("u" === c )){
next = 196;
}
break;
case 183:
if(("t" === c )){
next = 197;
}
break;
case 184:
if(("}" === c )){
next = 198;
}
break;
case 185:
if(("s" === c )){
next = 199;
}
break;
case 187:
if(("}" === c )){
next = 200;
}
break;
case 188:
if(("c" === c )){
next = 201;
}
break;
case 189:
if(("}" === c )){
next = 202;
}
break;
case 191:
if(("}" === c )){
next = 203;
}
break;
case 193:
if(("}" === c )){
next = 204;
}
break;
case 195:
if(("}" === c )){
next = 205;
}
break;
case 196:
if(("e" === c )){
next = 206;
}
break;
case 197:
if(("}" === c )){
next = 207;
}
break;
case 199:
if(("}" === c )){
next = 208;
}
break;
case 201:
if(("e" === c )){
next = 209;
}
break;
case 206:
if(("}" === c )){
next = 210;
}
break;
case 209:
if(("o" === c )){
next = 211;
}
break;
case 211:
if(("f" === c )){
next = 212;
}
break;
case 212:
if(("}" === c )){
next = 213;
}
break;
	}
	return next;
};

function CDFA_BLOCKCOMMENT(){
	this.ss=1;
	this.as=[2,3,4,5,6];
	this.tt=[null,null,72,72,72,72,71];
this.stt={};
}
CDFA_BLOCKCOMMENT.prototype= new CDFA_base();
CDFA_BLOCKCOMMENT.prototype.nextState = function(state, c){
    var next = 0;
    switch(state){
case 1:
if((c < "\n" || "\n" < c)  && (c < "\r" || "\r" < c)  && (c < "*" || "*" < c) ){
next = 2;
} else if(("\n" === c )){
next = 2;
} else if(("\r" === c )){
next = 2;
} else if(("*" === c )){
next = 5;
}
break;
case 5:
if(("/" === c )){
next = 6;
}
break;
	}
	return next;
};

var EOF={};
function Lexer(){

if(!(this instanceof Lexer)) return new Lexer();

this.pos={line:0,col:0};

this.states={};
this.state = ['DEFAULT'];
this.lastChar = '\n';
this.actions = [function anonymous() {

	return "boolean";

},function anonymous() {

	return "break";

},function anonymous() {

	return "byte";

},function anonymous() {

	return "case";

},function anonymous() {

	return "char";

},function anonymous() {

	return "class";

},function anonymous() {

	return "const";

},function anonymous() {

	return "continue";

},function anonymous() {

	return "default";

},function anonymous() {

	return "do";

},function anonymous() {

	return "double";

},function anonymous() {

	return "else";

},function anonymous() {

	return "extends";

},function anonymous() {

	return "float";

},function anonymous() {

	return "for";

},function anonymous() {

	return "if";

},function anonymous() {

	return "import";

},function anonymous() {

	return "instanceof";

},function anonymous() {

	return "int";

},function anonymous() {

	return "long";

},function anonymous() {

	return "new";

},function anonymous() {

	return "return";

},function anonymous() {

	return "short";

},function anonymous() {

	return "static";

},function anonymous() {

	return "super";

},function anonymous() {

	return "switch";

},function anonymous() {

	return "this";

},function anonymous() {

	return "void";

},function anonymous() {

	return "while";

},function anonymous() {

	this.jjval = parseFloat(this.jjtext);
	return "float";

},function anonymous() {

	this.jjval = parseInt(this.jjtext);
	return "integer";

},function anonymous() {

	return 'op_add';

},function anonymous() {

	return 'op_sub';

},function anonymous() {

	return 'op_mul';

},function anonymous() {

	return 'op_div';

},function anonymous() {

	return 'op_mod';

},function anonymous() {

	return 'op_addAssign';

},function anonymous() {

	return 'op_subAssign';

},function anonymous() {

	return 'op_mulAssign';

},function anonymous() {

	return 'op_divAssign';

},function anonymous() {

	return 'op_modAssign';

},function anonymous() {

	return 'op_equalCompare';

},function anonymous() {

	return 'op_LshiftEqual';

},function anonymous() {

	return 'op_RshiftEqual';

},function anonymous() {

	return 'op_greaterEqual';

},function anonymous() {

	return 'op_lessEqual';

},function anonymous() {

	return 'op_Lshift';

},function anonymous() {

	return 'op_Rshift';

},function anonymous() {

	return 'op_greater';

},function anonymous() {

	return 'op_less';

},function anonymous() {

	return 'op_assign';

},function anonymous() {

	return 'op_andand';

},function anonymous() {

	return 'op_oror';

},function anonymous() {

	return 'op_and';

},function anonymous() {

	return 'op_or';

},function anonymous() {

	return 'op_not';

},function anonymous() {

	return 'op_xor';

},function anonymous() {

	this.jjval = (this.jjtext == "true");
	return "boolean";

},function anonymous() {

	return "identifier";

},function anonymous() {

	return "string";

},function anonymous() {

	return "terminator";

},function anonymous() {

	return "field-invoker";

},function anonymous() {

	return "separator";

},function anonymous() {

	return "paranthesis-start";

},function anonymous() {

	return "paranthesis-end";

},function anonymous() {

	return "brackets-start";

},function anonymous() {

	return "brackets-end";

},function anonymous() {

	return "set-start";

},function anonymous() {

	return "set-end";

},function anonymous() {

	return "comment";

},function anonymous() {

	this.pushState('BLOCKCOMMENT');
	return "blockcomment-start";

},function anonymous() {

	this.popState();
	return "blockcomment-end";

},,function anonymous() {
 
},function anonymous() {
 console.log("EOF"); return "EOF"; 
}];
this.states["DEFAULT"] = {};
this.states["DEFAULT"].dfa = new CDFA_DEFAULT();
this.states["BLOCKCOMMENT"] = {};
this.states["BLOCKCOMMENT"].dfa = new CDFA_BLOCKCOMMENT();
}
Lexer.prototype.setInput=function (input){
        this.pos={row:0, col:0};
        if(typeof input === 'string')
        {input = new StringReader(input);}
        this.input = input;
        this.state = ['DEFAULT'];
        this.lastChar='\n';
        this.getDFA().reset();
        return this;
    };
Lexer.prototype.nextToken=function () {


        var ret = undefined;
        while(ret === undefined){
            this.resetToken();
            ret = this.more();
        }


        if (ret === EOF) {
            this.current = EOF;
        } else {
            this.current = {};
            this.current.name = ret;
            this.current.value = this.jjval;
            this.current.lexeme = this.jjtext;
            this.current.position = this.jjpos;
            this.current.pos = {col: this.jjcol, line: this.jjline};
        }
        return this.current;
    };
Lexer.prototype.resetToken=function (){
        this.getDFA().reset();
        this.getDFA().bol = (this.lastChar === '\n');
        this.lastValid = undefined;
        this.lastValidPos = -1;
        this.jjtext = '';
        this.remains = '';
        this.buffer = '';
        this.startpos = this.input.getPos();
        this.jjline = this.input.line;
        this.jjcol = this.input.col;
    };
Lexer.prototype.halt=function () {
        if (this.lastValidPos >= 0) {
            var lastValidLength = this.lastValidPos-this.startpos+1;
            this.jjtext = this.buffer.substring(0, lastValidLength);
            this.remains = this.buffer.substring(lastValidLength);
            this.jjval = this.jjtext;
            this.jjpos = this.lastValidPos + 1-this.jjtext.length;
            this.input.rollback(this.remains);
            var action = this.getAction(this.lastValid);
            if (typeof ( action) === 'function') {
                return action.call(this);
            }
            this.resetToken();
        }
        else if(!this.input.more()){//EOF
            var actionid = this.states[this.getState()].eofaction;
            if(actionid){
                action = this.getAction(actionid);
                if (typeof ( action) === 'function') {
                    //Note we don't care of returned token, must return 'EOF'
                    action.call(this);
                }
            }
            return EOF;
        } else {//Unexpected character
            throw new Error('Unexpected char \''+this.input.peek()+'\' at '+this.jjline +':'+this.jjcol);
        }
    };
Lexer.prototype.more=function (){
        var ret;
        while (this.input.more()) {
            var c = this.input.peek();
            this.getDFA().readSymbol(c);
            if (this.getDFA().isInDeadState()) {

                ret = this.halt();
                return ret;

            } else {
                if (this.getDFA().isAccepting()) {
                    this.lastValid = this.getDFA().getCurrentToken();
                    this.lastValidPos = this.input.getPos();

                }
                this.buffer = this.buffer + c;
                this.lastChar = c;
                this.input.next();
            }

        }
        ret = this.halt();
        return ret;
    };
Lexer.prototype.less=function (length){
        this.input.rollback(length);
    };
Lexer.prototype.getDFA=function (){
        return this.states[this.getState()].dfa;
    };
Lexer.prototype.getAction=function (i){
        return this.actions[i];
    };
Lexer.prototype.pushState=function (state){
        this.state.push(state);
        this.getDFA().reset();
    };
Lexer.prototype.popState=function (){
        if(this.state.length>1) {
            this.state.pop();
            this.getDFA().reset();
        }
    };
Lexer.prototype.getState=function (){
        return this.state[this.state.length-1];
    };
Lexer.prototype.restoreLookAhead=function (){
        this.tailLength = this.jjtext.length;
        this.popState();
        this.less(this.tailLength);
        this.jjtext = this.lawhole.substring(0,this.lawhole.length-this.tailLength);


    };
Lexer.prototype.evictTail=function (length){
        this.less(length);
        this.jjtext = this.jjtext.substring(0,this.jjtext.length-length);
    };
Lexer.prototype.isEOF=function (o){
        return o===EOF;
    }
;
function StringReader(str){
        if(!(this instanceof StringReader)) return new StringReader(str);
		this.str = str;
		this.pos = 0;
        this.line = 0;
        this.col = 0;
	}
StringReader.prototype.getPos=function (){
        return this.pos;
    };
StringReader.prototype.peek=function ()
	{
		//TODO: handle EOF
		return this.str.charAt(this.pos);
	};
StringReader.prototype.eat=function (str)
	{
		var istr = this.str.substring(this.pos,this.pos+str.length);
		if(istr===str){
			this.pos+=str.length;
            this.updatePos(str,1);
		} else {
			throw new Error('Expected "'+str+'", got "'+istr+'"!');
		}
	};
StringReader.prototype.updatePos=function (str,delta){
        for(var i=0;i<str.length;i++){
            if(str[i]=='\n'){
                this.col=0;
                this.line+=delta;
            }else{
                this.col+=delta;
            }
        }
    };
StringReader.prototype.rollback=function (str)
    {
        if(typeof str === 'string')
        {
            var istr = this.str.substring(this.pos-str.length,this.pos);
            if(istr===str){
                this.pos-=str.length;
                this.updatePos(str,-1);
            } else {
                throw new Error('Expected "'+str+'", got "'+istr+'"!');
            }
        } else {
            this.pos-=str;
            this.updatePos(str,-1);
        }

    };
StringReader.prototype.next=function ()
	{
		var s = this.str.charAt(this.pos);
		this.pos=this.pos+1;
		this.updatePos(s,1);
		return s;
	};
StringReader.prototype.more=function ()
	{
		return this.pos<this.str.length;
	};
StringReader.prototype.reset=function (){
        this.pos=0;
    };
if (typeof(module) !== 'undefined') { module.exports = Lexer; }
return Lexer;})();
